File: models/saturation.py
"""Saturation classification for GoTrends v2."""

from __future__ import annotations

from dataclasses import dataclass

import pandas as pd


@dataclass(frozen=True)
class SaturationConfig:
    high_impression_share: float = 0.90
    moderate_impression_share: float = 0.80
    high_lost_is_rank: float = 0.50
    critical_marginal_ratio: float = 0.70
    high_elasticity_floor: float = 0.35
    moderate_elasticity_floor: float = 0.70


def classify_saturation(row: pd.Series, config: SaturationConfig | None = None) -> tuple[str, str]:
    """Return saturation level and reason for one campaign row."""
    config = config or SaturationConfig()
    marginal_roas = row.get("marginal_roas")
    proxy_target_roas = row.get("proxy_target_roas")
    elasticity = row.get("elasticity")
    impression_share = row.get("impression_share")
    lost_is_rank = row.get("lost_is_rank")

    if pd.isna(marginal_roas) or pd.isna(proxy_target_roas):
        return "critical", "missing_marginal_or_proxy_target"
    if marginal_roas < proxy_target_roas * config.critical_marginal_ratio:
        return "critical", "marginal_roas_far_below_proxy_target"
    if pd.notna(elasticity) and elasticity < 0:
        return "critical", "negative_elasticity"
    if pd.notna(impression_share) and impression_share >= config.high_impression_share:
        return "high", "impression_share_above_90pct"
    if marginal_roas < proxy_target_roas:
        return "high", "marginal_roas_below_proxy_target"
    if pd.notna(elasticity) and elasticity < config.high_elasticity_floor:
        return "high", "low_elasticity"
    if pd.notna(impression_share) and impression_share >= config.moderate_impression_share:
        return "moderate", "impression_share_above_80pct"
    if pd.notna(lost_is_rank) and lost_is_rank >= config.high_lost_is_rank:
        return "moderate", "high_lost_is_rank"
    if pd.notna(elasticity) and elasticity < config.moderate_elasticity_floor:
        return "moderate", "moderate_elasticity"
    return "low", "room_to_scale"


def add_saturation_features(
    df: pd.DataFrame,
    config: SaturationConfig | None = None,
) -> pd.DataFrame:
    """Add saturation_level, reason, and pure budget block flag."""
    config = config or SaturationConfig()
    out = df.copy()

    if "proxy_target_roas" not in out.columns:
        out["proxy_target_roas"] = out.get("campaign_avg_roas").combine_first(
            out.get("campaign_type_avg_roas")
        )

    classifications = [classify_saturation(row, config) for _, row in out.iterrows()]
    out["saturation_level"] = [level for level, _ in classifications]
    out["saturation_reason"] = [reason for _, reason in classifications]
    out["pure_budget_increase_blocked"] = out["impression_share"].ge(
        config.high_impression_share
    )
    return out

File: app/src/models/saturation.ts
/**
 * Saturation classification for GoTrends v2.
 *
 * Port of legacy/python/models/saturation.py with 1e-6 parity.
 *
 * Adds three derived columns to each campaign row:
 *  - saturation_level: 'critical' | 'high' | 'moderate' | 'low'
 *  - saturation_reason: string explaining why
 *  - pure_budget_increase_blocked: impression_share >= high_impression_share
 */

export interface SaturationConfig {
  highImpressionShare: number
  moderateImpressionShare: number
  highLostIsRank: number
  criticalMarginalRatio: number
  highElasticityFloor: number
  moderateElasticityFloor: number
}

export const DEFAULT_SATURATION_CONFIG: SaturationConfig = {
  highImpressionShare: 0.90,
  moderateImpressionShare: 0.80,
  highLostIsRank: 0.50,
  criticalMarginalRatio: 0.70,
  highElasticityFloor: 0.35,
  moderateElasticityFloor: 0.70,
}

export type SaturationLevel = 'critical' | 'high' | 'moderate' | 'low'

export interface SaturationInputRow {
  proxy_target_roas?: number | null
  marginal_roas?: number | null
  elasticity?: number | null
  impression_share?: number | null
  lost_is_rank?: number | null
  campaign_avg_roas?: number | null
  campaign_type_avg_roas?: number | null
  [key: string]: unknown
}

export interface SaturationOutputRow extends SaturationInputRow {
  proxy_target_roas: number | null
  saturation_level: SaturationLevel
  saturation_reason: string
  pure_budget_increase_blocked: boolean
}

function isNum(v: unknown): v is number {
  return typeof v === 'number' && Number.isFinite(v)
}

/** Equivalent to pandas isna() check (covers null/undefined/NaN). */
function isNa(v: unknown): boolean {
  if (v === null || v === undefined) return true
  if (typeof v === 'number' && Number.isNaN(v)) return true
  return false
}

/** pandas.notna() */
function notNa(v: unknown): v is number {
  return isNum(v)
}

/** Returns [level, reason] for a single row. */
export function classifySaturation(
  row: SaturationInputRow,
  config: SaturationConfig = DEFAULT_SATURATION_CONFIG,
): [SaturationLevel, string] {
  const marginalRoas = row.marginal_roas
  const proxyTargetRoas = row.proxy_target_roas
  const elasticity = row.elasticity
  const impressionShare = row.impression_share
  const lostIsRank = row.lost_is_rank

  if (isNa(marginalRoas) || isNa(proxyTargetRoas)) {
    return ['critical', 'missing_marginal_or_proxy_target']
  }
  // After isNa guard, both are finite numbers.
  const m = marginalRoas as number
  const p = proxyTargetRoas as number

  if (m < p * config.criticalMarginalRatio) {
    return ['critical', 'marginal_roas_far_below_proxy_target']
  }
  if (notNa(elasticity) && elasticity < 0) {
    return ['critical', 'negative_elasticity']
  }
  if (notNa(impressionShare) && impressionShare >= config.highImpressionShare) {
    return ['high', 'impression_share_above_90pct']
  }
  if (m < p) {
    return ['high', 'marginal_roas_below_proxy_target']
  }
  if (notNa(elasticity) && elasticity < config.highElasticityFloor) {
    return ['high', 'low_elasticity']
  }
  if (notNa(impressionShare) && impressionShare >= config.moderateImpressionShare) {
    return ['moderate', 'impression_share_above_80pct']
  }
  if (notNa(lostIsRank) && lostIsRank >= config.highLostIsRank) {
    return ['moderate', 'high_lost_is_rank']
  }
  if (notNa(elasticity) && elasticity < config.moderateElasticityFloor) {
    return ['moderate', 'moderate_elasticity']
  }
  return ['low', 'room_to_scale']
}

/**
 * Adds saturation features to each row.
 *
 * Pandas semantics:
 *  - If `proxy_target_roas` is absent from the input, fallback to
 *    `campaign_avg_roas.combine_first(campaign_type_avg_roas)` (column-wise).
 *  - `pure_budget_increase_blocked = impression_share >= high_impression_share`
 *    Pandas Series.ge() treats NaN as False; we do the same.
 */
export function addSaturationFeatures<T extends SaturationInputRow>(
  rows: T[],
  config: SaturationConfig = DEFAULT_SATURATION_CONFIG,
): Array<T & {
  proxy_target_roas: number | null
  saturation_level: SaturationLevel
  saturation_reason: string
  pure_budget_increase_blocked: boolean
}> {
  if (rows.length === 0) return []

  const hasProxyCol = rows.some(r => 'proxy_target_roas' in r)

  return rows.map(r => {
    const next: Record<string, unknown> = { ...r }

    if (!hasProxyCol) {
      const avg = r.campaign_avg_roas
      const typeAvg = r.campaign_type_avg_roas
      next.proxy_target_roas = isNum(avg) ? avg : isNum(typeAvg) ? typeAvg : null
    } else if (next.proxy_target_roas === undefined) {
      next.proxy_target_roas = null
    }

    const [level, reason] = classifySaturation(next as SaturationInputRow, config)
    next.saturation_level = level
    next.saturation_reason = reason

    const is = (next as SaturationInputRow).impression_share
    next.pure_budget_increase_blocked = isNum(is) && is >= config.highImpressionShare

    return next as T & {
      proxy_target_roas: number | null
      saturation_level: SaturationLevel
      saturation_reason: string
      pure_budget_increase_blocked: boolean
    }
  })
}
